### PR TITLE
Add `PackedIterableDataset`, `peft_module_casting_to_bf16` util method and `append_concat_token` flag

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. Fixes https://github.com/huggingface/trl/issues/746 by introducing `PackedIterableDataset`
2. Inline with QLoRA code when using bf16 [here](https://github.com/artidoro/qlora/blob/main/qlora.py#L396-L405), added `peft_module_casting_to_bf16`.
3. Added `append_concat_token` flag so that datasets which already have eos_token at the end of each sample can skip appending the eos/concat token.

Details:
Issues with `ConstantLengthDataset`: 
1. Learning rate schedulers like Cosine, linear_decay ... will run improperly because the length of the dataset is set to the initial dataset while an epoch will have much fewer steps because of packing.
For example, I've the below code which uses the existing `ConstantLengthDataset` and epoch=1. The logs show that epoch only has 207 steps instead of the original 1360 steps due to packing. Because of this, cosine lr_scheduler doesn't go down to 0 at the end of the epoch.
Code: https://github.com/pacman100/DHS-LLM-Workshop/pull/14
Commnad:
```
cd /DHS-LLM-Workshop/chat_assistant/training

accelerate launch --config_file configs/fsdp_config.yaml \
train.py \
--model_name_or_path "meta-llama/Llama-2-7b-chat-hf" \
--dataset_name "smangrul/code-chat-assistant-v1" \
--max_seq_len 4096 \
--num_train_epochs 1 \
--save_strategy "epoch" \
--logging_steps 25 \
--eval_steps 100 \
--bf16 True \
--packing True \
--output_dir "experiments/full-finetune-llama7B-fullshard" \
--per_device_train_batch_size 1 \
--gradient_accumulation_steps 1 \
--dataset_text_field "content" \
--learning_rate 5e-5  \
--lr_scheduler_type "cosine" \
--weight_decay 0.01 \
--warmup_ratio 0.03 \
--use_flash_attn True
```

Output logs:
```
15%|█▌        | 207/1360 [04:09<22:52,  1.19s/it]
```

wandb_plots:
Screenshot 2023-11-23 at 3 17 58 PM

So, epoch based training using `ConstantLengthDataset` is not preferable with decay based lr_scheduler and packing=True.

2. When `max_steps` is passed which has more steps than the initial `len(dataset)`, it results in https://github.com/huggingface/trl/issues/746. `PackedIterableDataset` is meant to fix this issue as it doesn't contain the length attribute and works as expected when passing `max_steps`.

### CODE REVIEW

**Comment by pacman100** [comment]
cc @younesbelkada, @lvwerra

**Comment by pacman100** [comment]
Closing this in favour of https://github.com/huggingface/trl/pull/1110
